### PR TITLE
Add `queryParams()` to `ServiceRequestContext`

### LINKED ISSUE #3955: Add `queryParams()` to `ServiceRequestContext`
It would be convenient to access query parameters from `ServiceRequestContext` like `pathParams`.
https://github.com/line/armeria/blob/bcb621e46c26f0ac84b4b820687383a99b509019/core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java#L251-L259
```java
interface ServiceRequestContext {
    QueryParams queryParams();
    String queryParam(String name);
}
```

### PR DESCRIPTION
**Pull request by injae-kim**
**Related Issue** #3955

### Motivation:

Currently, we support `pathParams`, `pathParam` on `ServiceRequestContext`.
So it seems good to support `queryParams`, `queryParam` too.

### Modifications:

- Add `queryParams`, `queryParam` on `ServiceRequestContext`.

### Result:

- You can now use `queryParams` to get `QueryParams` on `ServiceRequestContext`.
- Close #3955 issue

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java [hunk 1]
```diff
@@ -258,6 +260,28 @@ default String pathParam(String name) {
...
+        return queryParams().get(name);
+    }
+
+    /**
+     * Returns the values of the specified query parameter.
+     */
+    @Nullable
+    default List<String> queryParams(String name) {
+        return queryParams().getAll(name);
+    }
```
Comment (injae-kim) [inline core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java:283]: I found that `QueryParams` also support `List<String> getAll(name)` so added this method too!

WDYT? If this method is unnecessary, then I'll remove it!

File: core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java [hunk 2]
```diff
@@ -258,6 +260,28 @@ default String pathParam(String name) {
...
     }
 
+    /**
+     * Returns the query parameters mapped by the {@link Route} associated with the {@link Service}
+     * that is handling the current {@link Request}.
+     */
+    QueryParams queryParams();
+
+    /**
+     * Returns the values of the specified query parameter.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java:270]: Could you copy/paste the Javadoc of `QueryParamGetters`?

File: core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java [hunk 3]
```diff
@@ -258,6 +260,28 @@ default String pathParam(String name) {
...
+     */
+    @Nullable
+    default List<String> queryParams(String name) {
+        return queryParams().getAll(name);
+    }
+
+    /**
+     * Returns the value of the specified query parameter.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java:278]: ditto. It's hard to notice the difference between this description and the description of the above method. 😉

**Review by minwoox — COMMENTED** [review]
Nice work!
Left a minor comment. 😉

File: core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java (hunk 1 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java]: I think it's okay. 😄

File: core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java [hunk 4]
```diff
@@ -258,6 +260,28 @@ default String pathParam(String name) {
         return pathParams().get(name);
     }
 
+    /**
+     * Returns the query parameters mapped by the {@link Route} associated with the {@link Service}
+     * that is handling the current {@link Request}.
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java:265]: Unlike `pathParams`, query parameters are not heavily used by `Route`. Although it is possible.
Could we explain where/how this value came from? 
```
Returns the decoded {@linkplain QueryParams query part} of the current {@link Request} URI,
as defined in <a href="https://datatracker.ietf.org/doc/rfc3986/">RFC3986</a>.
```

File: core/src/test/java/com/linecorp/armeria/server/ServiceRequestContextTest.java [hunk 5]
```diff
@@ -167,6 +171,38 @@ void pushWithOldIrrelevantServiceCtx() {
         assertCurrentCtx(null);
     }
 
+    @Test
+    void queryParams() {
+        final QueryParams queryParams = QueryParams.of("param1", "value1",
```
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/server/ServiceRequestContextTest.java:176]: I couldn't find the relevance between `queryParams` and `ctx.queryParams()` before I look into `serviceRequestContextWithPathAndQuery`. (I also had to carefully take a look at the parameter if there's something different.)

I think it's better to pass the `queryParams` to the method to create the `ctx` or
we can just inline the `serviceRequestContextWithPathAndQuery()` method and merge those three test cases added. WDYT?

Comment (injae-kim) [reply core/src/test/java/com/linecorp/armeria/server/ServiceRequestContextTest.java:176]: > we can just inline the serviceRequestContextWithPathAndQuery() method and merge those three test cases added. WDYT?

I think this way looks much better! 👍  I'll update it!

**Review by ikhoon — APPROVED** [review]
Thanks, @injae-kim! 💯

**Review by minwoox — APPROVED** [review]
Awesome. 😆

**Review by jrhee17 — APPROVED** [review]
💯 I think this is a useful addition that would be used widely. Thank you @injae-kim! 🙇

File: core/src/test/java/com/linecorp/armeria/server/ServiceRequestContextTest.java [hunk 6]
```diff
@@ -167,6 +171,30 @@ void pushWithOldIrrelevantServiceCtx() {
...
+
+        assertThat(ctx.queryParam("param1")).isEqualTo("value1");
+        assertThat(ctx.queryParam("Param1")).isEqualTo("Value3");
+        assertThat(ctx.queryParam("PARAM1")).isEqualTo("VALUE4");
+        assertThat(ctx.queryParam("Not exist")).isNull();
+
+        assertThat(ctx.queryParams("param1")).isEqualTo(ImmutableList.of("value1", "value2"));
+        assertThat(ctx.queryParams("Param1")).isEqualTo(ImmutableList.of("Value3"));
+        assertThat(ctx.queryParams("PARAM1")).isEqualTo(ImmutableList.of("VALUE4"));
+        assertThat(ctx.queryParams("Not exist")).isEqualTo(Collections.emptyList());
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/server/ServiceRequestContextTest.java:195]: nit:
```suggestion
        assertThat(ctx.queryParams("Not exist")).isEmpty();
```

Comment (injae-kim) [reply core/src/test/java/com/linecorp/armeria/server/ServiceRequestContextTest.java]: oh, thanks for detailed review 🙇‍♂️

**Comment by injae-kim** [comment]
Thanks to all reviewers about detailed review! 🙇‍♂️

**Comment by minwoox** [comment]
Congratulations on your first PR! 🎉 🎉 🎉

**Pull request merged**